ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -590,7 +590,7 @@ void fillChunks() {
         // the chunk above.
         PutChunk lastChunk = getBuildingChunk();
         if (lastChunk != null) {
-          if (chunkCounter != 0 && lastChunk.buf.readableBytes() == 0) {
+          if (chunkCounter != 0 && (lastChunk.buf == null || lastChunk.buf.readableBytes() == 0)) {
```

Comment:
is there a unit test we could write to reproduce the NPE?

Reply:
Unfortunately, I haven't figured out the root cause of this exception. But it keeps happening, so I want to push out a fix first.

Reply:
In what scenario lastChunk != null but lastChunk == null ?

Reply:
The root cause is found and the test cases added.